api/src/main/java/org/openmrs/module/reporting/report/renderer/DelimitedTextReportRenderer.java

```diff
@@ -234,44 +235,53 @@ void writeDataSet(DataSet dataset, OutputStream out, String textDelimiter, Strin
 		w.write(lineEnding);
 
 		// data rows
-		for (DataSetRow row : dataset) {
-			for (Iterator<DataSetColumn> i = columns.iterator(); i.hasNext();) {
-				DataSetColumn column = i.next();
-				Object colValue = row.getColumnValue(column);
-				w.write(textDelimiter);
-				if (colValue != null) {
-                    String toPrint;
-					if (colValue instanceof Cohort) {
-                        toPrint = escape(Integer.toString(((Cohort) colValue).size()));
-					}
-					else if (colValue instanceof IndicatorResult) {
-                        toPrint = ((IndicatorResult) colValue).getValue().toString();
-					}
-					else if (dateFormat != null && colValue instanceof Date) {
-                        toPrint = dateFormat.format((Date)colValue);
-                    }
-					else {
-						// this check is because a logic EmptyResult .toString() -> null
-						String temp = escape(colValue.toString());
-						if (temp != null) {
-                            toPrint = temp;
-                        } else {
-                            toPrint = "";
-                        }
-					}
-                    w.write(filterBlacklist(toPrint, blacklist));
-				}
-				w.write(textDelimiter);
-				if (i.hasNext()) {
-					w.write(fieldDelimiter);
-				}
-			}
-			w.write(lineEnding);
-		}
-
+        if (dataset instanceof IterableSqlDataSet) {

```

Comment:
I'm curious here to see if we really need to do this check, or if we can use the fact that all datasets have the ability to iterate across the data set rows to mean that we don't need custom logic here, but can leave the nature of the iteration to the internals of the data set implementation.

Reply:
I tried that.
The problem is that the `Iterators` have different behaviors when they have no more elements.
- The `SqlIterator` will return null when it have no more elements.
- The `Iterator` throws a `NoSuchElementException` when it have no more records.

I would have to do something like:
```
try {
    row = sqlIterator.next();
} catch (NoSuchElementException ex) {
    row = null;
}
```

Reply:
But wouldn't you do this in a while or for loop, checking on iterator.hasNext() before calling iterator.next() ?

Reply:
That was my first approach, but @dkayiwa pointed that it may have some flaws:
See the [discussion](https://github.com/openmrs/openmrs-module-reporting/pull/213#discussion_r603443586):
> Can you also take a look at this? https://rules.sonarsource.com/java/tag/performance/RSPEC-2232

Reply:
I see.  OK.  There are probably other solutions too, if you think they are better.  For example, keeping "current" and "next" as state within the iterator, and updating / returning these as needed when "next" and "hasNext" are called.

Reply:
In that case, the `SqlIterator.hasNext()` would call the `Iterator.next()` keep it in memory and return it when the `SqlIterator.next()` is call, is that it?
That works for me.

The problem with that is if someone try to only use `SqlIterator.next()`, it will always return the same value. we can try to solve that calling ourselves the `Iterator.next()` if it is called more than once. 
It's not clean but it should work, let me know if this is OK for you?

Reply:
Well, I think you'd not want to expose the actual SqlIterator for arbitrary use, but only allow access through the Data Set API methods

Reply:
But the `DataSet` already give you access to the Iterator.
I have made the changes to have it working with the `hasNext()`. It's a little messy but it does the job.
Let me know if its OK for you.

Reply:
Sure - I don't want to get too bogged down in this - I was just trying to make suggestions, not trying to be a blocker.  If the entire PR is ready for review again, let me know.

Reply:
Yes, I think is ready for another review :)